core/distributed/matrix.cpp

```diff
@@ -0,0 +1,440 @@
+    this->set_size({});
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Matrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
+    const device_matrix_data<value_type, global_index_type>& data,
+    const Partition<local_index_type, global_index_type>* row_partition,
+    const Partition<local_index_type, global_index_type>* col_partition)
+{
+    const auto comm = this->get_communicator();
+    GKO_ASSERT_EQ(data.get_size()[0], row_partition->get_size());
+    GKO_ASSERT_EQ(data.get_size()[1], col_partition->get_size());
+    GKO_ASSERT_EQ(comm.size(), row_partition->get_num_parts());
+    GKO_ASSERT_EQ(comm.size(), col_partition->get_num_parts());
+    auto exec = this->get_executor();
+    auto local_part = comm.rank();
+
+    // set up LinOp sizes
+    auto num_parts = static_cast<size_type>(row_partition->get_num_parts());
+    auto global_num_rows = row_partition->get_size();
+    auto global_num_cols = col_partition->get_size();
+    dim<2> global_dim{global_num_rows, global_num_cols};
+    this->set_size(global_dim);
+
+    // temporary storage for the output
+    array<local_index_type> diag_row_idxs{exec};
+    array<local_index_type> diag_col_idxs{exec};
+    array<value_type> diag_values{exec};
+    array<local_index_type> offdiag_row_idxs{exec};
+    array<local_index_type> offdiag_col_idxs{exec};
+    array<value_type> offdiag_values{exec};
+    array<local_index_type> recv_gather_idxs{exec};
+    array<comm_index_type> recv_offsets_array{exec, num_parts + 1};
+
+    // build diagonal, off-diagonal matrix and communication structures
+    exec->run(matrix::make_build_diag_offdiag(
+        data, make_temporary_clone(exec, row_partition).get(),
+        make_temporary_clone(exec, col_partition).get(), local_part,
+        diag_row_idxs, diag_col_idxs, diag_values, offdiag_row_idxs,
+        offdiag_col_idxs, offdiag_values, recv_gather_idxs,
+        recv_offsets_array.get_data(), non_local_to_global_));
+
+    // read the local matrix data
+    const auto num_diag_rows =
+        static_cast<size_type>(row_partition->get_part_size(local_part));
+    const auto num_diag_cols =
+        static_cast<size_type>(col_partition->get_part_size(local_part));
+    const auto num_ghost_cols = non_local_to_global_.get_num_elems();
+    device_matrix_data<value_type, local_index_type> diag_data{
+        exec, dim<2>{num_diag_rows, num_diag_cols}, std::move(diag_row_idxs),
+        std::move(diag_col_idxs), std::move(diag_values)};
+    device_matrix_data<value_type, local_index_type> offdiag_data{
+        exec, dim<2>{num_diag_rows, num_ghost_cols},
+        std::move(offdiag_row_idxs), std::move(offdiag_col_idxs),
+        std::move(offdiag_values)};
+    as<ReadableFromMatrixData<ValueType, LocalIndexType>>(this->local_mtx_)
+        ->read(diag_data);
+    as<ReadableFromMatrixData<ValueType, LocalIndexType>>(this->non_local_mtx_)
+        ->read(offdiag_data);
```

Comment:
If the requested type is COO, then you can just move the arrays into the COO object, right ?